Extend -o to work when compiling C files

shindere opened the pull request:
This PR fixes issue #7677.

When using ocamlc/ocamlopt to compile C files, it is currently not possible
to control where the C compiler should place the resulting object file.

In other words, without this PR, a command like

```shell
ocamlopt -c foo.c -o /tmp/foo.o
```

produced the following error message:

Options -c and -o are incompatible when compiling C files

This PR removes this restriction and lets the compiler accept
that the command-line options -c and -o are given simultaneously
also when compiling one C file, so that the previous command now works.

The -o command-line option remains however forbidden when compiling
several C files during the same compiler invocation.

The PR includes tests to ensure the compiler behaves as expected when
only one C file is provided on the command-line and when several
of them are provided.

A `file-exists` action has been added to ocamltest to make testing for the
presence of the object file easier.

The code has been tested with precheck, see build
[##521](https://ci.inria.fr/ocaml/job/precheck/521/).

Cc @dbuenzli

Octachron commented on driver/compenv.ml:
```diff
@@ -656,12 +661,12 @@ let process_deferred_actions env =
+          if List.length (List.filter (function
+              | ProcessCFile _ -> true
+              | _ -> false) !deferred_actions) > 1 then
+            fatal "Option -o makes no sense when compiling several C files";
```
Shouldn't this case should be combined with the one below to reject
```ocaml
ocamlopt -c a.c b.ml -o something
```
?

shindere commented:
Florian Angeletti (2020/10/13 01:16 -0700):
> Shouldn't this case should be combined with the one below to reject
> ```ocaml
> ocamlopt -c a.c b.ml -o something
> ```
> ?

That was definitely a good idea, many thanks for the suggestion.

The PR has been rebased on latest trunk and updated accordingly.

shindere commented:
Is there anything that prevents this from being approved / merged?

Would be happy to fix any remaining issue.

gasche reviewed (approved):
The code looks good and the feature is useful.

The change needs a Changes entry.

shindere commented:
Many thanks for your prompt response and approval, @gasche.

The PR has been rebased on latest trunk and a Changes entry added.

gasche commented:
Thanks! `check-typo` complains:

```
Checking 6f76c388bee0b478dbbb2c2d82e46b389fae66f8: ocamltest/builtin_actions.ml
./ocamltest/builtin_actions.ml:224.35: [white-at-eol] whitespace at end of line
./ocamltest/builtin_actions.ml:226.15: [white-at-eol] whitespace at end of line
```

shindere commented:
Gabriel Scherer (2021/01/05 04:18 -0800):
> Thanks! `check-typo` complains:
[...]

Gosh. Sorry, and thanks. Fixed.

gasche commented:
`check-typo` is still complaining in the same way, are you sure that you pushed your fixing change?

shindere committed:
Extend -o to work when compiling C files

When using ocamlc/ocamlopt to compile C files, it was not possible
so far to control where the resulting object file should be written.

In other words, before this commit a command like

ocamlopt -c foo.c -o /tmp/foo.o

produced the following error message:

Options -c and -o are incompatible when compiling C files

This commit removes this restriction and lets the compiler accept
that the command-line options -c and -o are given simultaneously
also when compiling one C file, so that the previous command now works.

The -o command-line option remains forbidden when compiling several source
files.

shindere commented:
Gabriel Scherer (2021/01/05 08:48 -0800):
> `check-typo` is still complaining in the same way, are you sure that
> you pushed your fixing change?

They should be there now. Sorry.

shindere commented:
Gabriel Scherer (2021/01/06 00:07 -0800):
> Merged #9960 into trunk.

Great! Thanks!

nojb commented:
There are two issues with this patch:

1. It broke previous behaviour of `ocaml` when you *don't* pass `-o`:
```
(before) ocamlc foo.c --> calls "$cc -c foo.c"
(after) ocamlc foo.c --> calls "$cc -o foo.o -c foo.c"
```
This causes problems when people use:
```
ocamlc -ccopt '-o whatever' foo.c
```
which was necessary before this PR.

2. The `-o` flag to `ocaml{c,opt}` is ignored if `-c` is not passed as well when compiling C files.

ghost commented:
cf. https://github.com/moosotc/llpp/issues/174
